File: frontend/app/utils/home-formatters.ts
import type { HomeCurrency } from "@/app/types/home";

const rubFormatter = new Intl.NumberFormat("ru-RU", {
  maximumFractionDigits: 0,
});

const usdFormatter = new Intl.NumberFormat("ru-RU", {
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
});

const eurFormatter = new Intl.NumberFormat("ru-RU", {
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
});

const nextPaymentDateFormatter = new Intl.DateTimeFormat("ru-RU", {
  day: "numeric",
  month: "short",
});

const toValidDate = (value: Date | string | null | undefined): Date | null => {
  if (!value) {
    return null;
  }

  const date = value instanceof Date ? value : new Date(value);

  if (Number.isNaN(date.getTime())) {
    return null;
  }

  return date;
};

export const formatMoney = (value: number, currency: HomeCurrency) => {
  if (currency === "usd") {
    return `$${usdFormatter.format(value)}`;
  }

  if (currency === "eur") {
    return `€${eurFormatter.format(value)}`;
  }

  return `${rubFormatter.format(Math.round(value))} ₽`;
};

export const formatNextPayment = (value: Date | string | null | undefined) => {
  const parsedDate = toValidDate(value);

  if (!parsedDate) {
    return "дата списания не указана";
  }

  return `следующий платёж ${nextPaymentDateFormatter.format(parsedDate)}`;
};

export const formatNextPaymentShort = (
  value: Date | string | null | undefined,
) => {
  const parsedDate = toValidDate(value);

  if (!parsedDate) {
    return "Не указана";
  }

  return nextPaymentDateFormatter.format(parsedDate);
};

export const formatSubscriptionCount = (count: number) => {
  const mod10 = count % 10;
  const mod100 = count % 100;

  if (mod10 === 1 && mod100 !== 11) {
    return `${count} активная подписка`;
  }

  if (mod10 >= 2 && mod10 <= 4 && !(mod100 >= 12 && mod100 <= 14)) {
    return `${count} активные подписки`;
  }

  return `${count} активных подписок`;
};

File: mobile/lib/features/home/presentation/home_formatters.dart
import 'package:intl/intl.dart';

final NumberFormat _rubFormatter = NumberFormat('#,##0', 'ru_RU');
final NumberFormat _usdFormatter = NumberFormat('#,##0.00', 'ru_RU');
final NumberFormat _eurFormatter = NumberFormat('#,##0.00', 'ru_RU');
final DateFormat _nextPaymentDateFormatter = DateFormat('d MMM', 'ru_RU');

String formatMoney(num value, String currency) {
  if (currency == 'usd') {
    return '\$${_usdFormatter.format(value)}';
  }

  if (currency == 'eur') {
    return '€${_eurFormatter.format(value)}';
  }

  final rounded = value.round();
  return '${_rubFormatter.format(rounded)} ₽';
}

String formatRub(num value) {
  return formatMoney(value, 'rub');
}

String formatNextPayment(DateTime? value) {
  if (value == null) {
    return 'дата списания не указана';
  }

  return 'следующий платёж ${_nextPaymentDateFormatter.format(value)}';
}

String formatNextPaymentShort(DateTime? value) {
  if (value == null) {
    return 'Не указана';
  }

  return _nextPaymentDateFormatter.format(value);
}

String formatSubscriptionCount(int count) {
  final mod10 = count % 10;
  final mod100 = count % 100;

  if (mod10 == 1 && mod100 != 11) {
    return '$count активная подписка';
  }
  if (mod10 >= 2 && mod10 <= 4 && !(mod100 >= 12 && mod100 <= 14)) {
    return '$count активные подписки';
  }

  return '$count активных подписок';
}

String formatPeriodLabel(int period) {
  final safePeriod = period < 1 ? 1 : period;

  if (safePeriod == 1) {
    return 'Ежемесячно';
  }
  if (safePeriod == 12) {
    return 'Раз в год';
  }

  return 'Раз в $safePeriod ${_monthsWord(safePeriod)}';
}

String currencyLabel(String currency) {
  if (currency == 'usd') {
    return 'USD';
  }

  if (currency == 'eur') {
    return 'EUR';
  }

  return 'RUB';
}

String _monthsWord(int value) {
  final mod10 = value % 10;
  final mod100 = value % 100;

  if (mod10 == 1 && mod100 != 11) {
    return 'месяц';
  }
  if (mod10 >= 2 && mod10 <= 4 && !(mod100 >= 12 && mod100 <= 14)) {
    return 'месяца';
  }

  return 'месяцев';
}
